Fix issue with convert function when converting string to date, datetimeoffset, datetime2, datetime, smalldatetime or time

rohit01010 opened the pull request:
### Description
This PR will add fix for following issues:
- For converting string to datetime2, datetimeoffset and smalldatetime using convert function, we are directly using cast function which will convert the string to given type with default style. Added support to use corresponding SQL helper functions defined which will fix this issue.
- For conversion of string to date, datetime, smalldatetime and time using convert function, there are helper functions defined but are unable to handle all the supported datetime string literals. Refactored these definitions and fixed this issue.
- For conversion of string to datetime2 and datetimeoffset using convert function, there are no helper functions defined which handles this conversion as there are for conversion of string to date, datetime, smalldatetime and time. Added definitions to handle these datatypes as well.
- Function sys.babelfish_get_microsecs_from_fractsecs() is unable to handle the edge case where fractional seconds get overflowed after rounding it to given scale.
- Function babelfish_get_timeunit_from_string only returns HOURS, MINUTES, SECONDS and FRACTSECONDS from a given time string. Extended this function to return OFFHOURS, OFFMINUTES and OFFSIGN to get details of timeoffset from time string.

First commit changes cherry-picked from: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2613

Authored-by: Rohit Bhagat [rohitbgt@amazon.com](mailto:rohitbgt@amazon.com)
Signed-off-by: Rohit Bhagat [rohitbgt@amazon.com](mailto:rohitbgt@amazon.com)

### Issues Resolved
BABEL-4896

### Test Scenarios Covered ###
* **Use case based -** YES


* **Boundary conditions -** YES


* **Arbitrary inputs -** YES


* **Negative test cases -** YES


* **Minor version upgrade tests -** YES


* **Major version upgrade tests -** YES


* **Performance tests -** NA


* **Tooling impact -** NA


* **Client tests -** YES



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

rishabhtanwar29 commented on contrib/babelfishpg_tsql/sql/sys_function_helpers.sql:
```diff
@@ -389,6 +389,12 @@ BEGIN
     ELSE
         v_fseconds := sys.babelfish_get_microsecs_from_fractsecs(to_char(v_datetimeval, 'US'), v_scale);
 
+        -- Following condition will handle for overflow of fractsecs
```
Nit: `Following condition will handle overflow of fractsecs`

rohit01010 replied:
Done

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -196,11 +196,60 @@ TsqlFunctionConvert(TypeName *typename, Node *arg, Node *style, bool try, int lo
+		Node	   *helperFuncCall;
+		helperFuncCall = (Node *) makeFuncCall(TsqlSystemFuncName("babelfish_conv_helper_to_time"), lcons(makeIntConst(typmod, location), args), COERCE_EXPLICIT_CALL, location);
+		
+		/*
+		 * add a type cast on top of the CONVERT helper function
+		 * so typmod can be applied
+		 */
```
Can be re-written as: `apply typmod cast on top of CONVERT`

rohit01010 replied:
existing comment looks simple and easy to understand, let's not change it.

rishabhtanwar29 commented on test/JDBC/upgrade/16_2/schedule:
```diff
@@ -527,4 +527,10 @@ babel_test_int4_numeric_oper_before_16_3
 babel_test_int8_numeric_oper_before_16_3
 babel_test_int2_numeric_oper_before_16_3
 BABEL_3571-15_5-or-15_6-or-16_1-or-16_2
+test_conv_string_to_date-before-16_3
+test_conv_string_to_datetime
+test_conv_string_to_datetime2-before-16_3
+test_conv_string_to_datetimeoffset-before-16_3
+test_conv_string_to_smalldatetime
+test_conv_string_to_time-before-16_3
```
Why have you only added these upto 16.2, 15.6 and 14.11? Why not add these all previous versions?

rohit01010 replied:
Added in all the base versions

rishabhtanwar29 commented on test/JDBC/upgrade/15_7/schedule:
```diff
@@ -516,4 +516,10 @@ babel_test_int8_numeric_oper
 babel_test_int2_numeric_oper
 BABEL_3571
 test_like_for_AI
+test_conv_string_to_date-before-16_3
+test_conv_string_to_datetime
+test_conv_string_to_datetime2-before-16_3
+test_conv_string_to_datetimeoffset-before-16_3
+test_conv_string_to_smalldatetime
+test_conv_string_to_time-before-16_3
```
We are going to backpatch this to 15 and 14 as well so then ideally we shouldn't be using `-before` suffix in 15_7 and 14_12 schedule files but it will fail until we actually merge this change to those versions. So, let's make sure to have PRs for the previous versions merge them from 14 to 16 order.

rohit01010 replied:
yes, we will merge this PR, once the PR's from previous version's gets merged

rohit01010 committed:
Fix issue with conversion of string to date related types

Signed-off-by: Rohit Bhagat <rohitbgt@amazon.com>

rohit01010 committed:
renamed and updated test files and made some changes related for 4X

Signed-off-by: Rohit Bhagat <rohitbgt@amazon.com>

rohit01010 committed:
used new definition of conv_string_to_datetime in date_bucket function

Signed-off-by: Rohit Bhagat <rohitbgt@amazon.com>

rohit01010 committed:
fix issue in upgrade from 14.latest to 15.latest and 15.latest to 16.latest

Signed-off-by: Rohit Bhagat <rohitbgt@amazon.com>

rohit01010 committed:
Revert "fix issue in upgrade from 14.latest to 15.latest and 15.latest to 16.latest"

This reverts commit 6dfa96da12ccdaf127f9f85bd29fafb3a7746f31.

rohit01010 committed:
renamed deprecated functions and updated schedule file

Signed-off-by: Rohit Bhagat <rohitbgt@amazon.com>

rohit01010 committed:
retrigger workflows

Signed-off-by: Rohit Bhagat <rohitbgt@amazon.com>

rohit01010 committed:
removed a testcase as it was giving different output for upgrade path with same source and target version

This testcase which was removed was returning different output in following two upgrade paths:-
* 15.5 to 15.7 to 16.latest
* 15.5 to 16.latest
and since we cannot have different expected output for these two upgrade paths, as of now we had removed it

Signed-off-by: Rohit Bhagat <rohitbgt@amazon.com>

rohit01010 committed:
retrigger workflows

Signed-off-by: Rohit Bhagat <rohitbgt@amazon.com>

rohit01010 committed:
Merge branch 'BABEL_4_X_DEV' into BABEL-4896
